tests/objects/test_generic_objects.py

```diff
@@ -136,15 +136,26 @@ class Value(Generic[T]):
     class Foo:
         string: Value[str]
         strings: Value[List[str]]
+        optional_string: Value[Optional[str]]
+        optional_strings: Value[Optional[List[str]]]
 
     definition = Foo._type_definition
     assert definition.name == "Foo"
     assert not definition.is_generic
-    [string_field, strings_field] = definition.fields
+    [
+        string_field,
+        strings_field,
+        optional_string_field,
+        optional_strings_field,
+    ] = definition.fields
     assert string_field.python_name == "string"
     assert string_field.type._type_definition.name == "StrValue"
     assert strings_field.python_name == "strings"
     assert strings_field.type._type_definition.name == "StrListValue"
+    assert optional_string_field.python_name == "optional_string"
+    assert optional_string_field.type._type_definition.name == "StrOptionalValue"
+    assert optional_strings_field.python_name == "optional_strings"
+    assert optional_strings_field.type._type_definition.name == "StrListOptionalValue"
```

Comment:
The name of this is weird. A human would probably call it a `OptionalStrListValue` but generating it this way could introduce ambiguous type names. e.g.`Optional[Value[List[str]]]` and `Value[List[Optional[str]]]` would result in the same type name being generated.

Reply:
This could potentially be solved by having the name generate inside->out instead of outside->in, _except_ for the outermost element, which would go at the end. This would result in:
- `Value[str]` -> `StrValue`
- `Value[List[str]]` -> `ListStrValue`
- `Value[Optional[str]]` -> `OptionalStrValue`
- `Value[Optional[List[str]]]` -> `OptionalListStrValue`

Reply:
I'd rather leave it all `inside->out` or all `outside->in` as you already show some weird behavior otherwise. Why is a `Value` of `StrList` called `ListStrValue`. That doesn't make much sense.

tbh. maybe it makes sense to change this to a all `inside->out` name generation in the future?

- Value[str] -> ValueStr
- Value[List[str]] -> ValueListStr
- Value[Optional[str]] -> ValueOptionalStr
- Value[Optional[List[str]]] -> ValueOptionalListStr

Reply:
> Why is a Value of `StrList` called `ListStrValue`

I think that makes sense, actually. In English, adjectives usually go before the noun. Consider "green tree". It's a `tree` of type `green`. Same with `StrListValue`. It's a `Value` of type `StrList`.

Reply:
`ListStrValue != `StrListValue`

I had to read your comment several times to see your mistake. You misread my comment.

I guess we can all agree that either  all `inside->out` or `outside->in` makes sense. Mixing those two just gets confusing a lot as you just showed as well. :grinning: :+1:

Reply:
I'd say let's do this:

```
Value[str] -> ValueStr
Value[List[str]] -> ValueListStr
Value[Optional[str]] -> ValueOptionalStr
Value[Optional[List[str]]] -> ValueOptionalListStr
```

as you suggested @bikeshedder 

I'll open an issue later to allow us to override the names for generic (maybe using Annotated).

Let me know if you can do this @bikeshedder, otherwise I'll try to do it tonight (I'm on holidays so I'm not using my laptop often :))

Reply:
ok, did the change, I want to get @jkimbo's and @AndrewIngram's opinion too, as this is now a breaking change as it changed all the concrete types made from generics, which might be annoying for people using generics.

Generics aren't documented yet, but this might still affect quite a few people

Reply:
I'm happy with both styles. Though I have to agree that the `outside->in` order makes it way easier to reason about since the generic reads the same as the generated name just leaving out the square brackets.

Reply:
I think this makes sense @patrick91 👍